provider/cloudflare.go

```diff
@@ -119,29 +126,33 @@ func NewCloudFlareProvider(domainFilter DomainFilter, zoneIDFilter ZoneIDFilter,
 		zoneIDFilter: zoneIDFilter,
 		proxied:      proxied,
 		DryRun:       dryRun,
+		PaginationOptions: cloudflare.PaginationOptions{
+			PerPage: zonesPerPage,
+		},
 	}
 	return provider, nil
 }
 
 // Zones returns the list of hosted zones.
 func (p *CloudFlareProvider) Zones() ([]cloudflare.Zone, error) {
 	result := []cloudflare.Zone{}
-
-	zones, err := p.Client.ListZones()
+	ctx := context.TODO()
+	zonesResponse, err := p.Client.ListZonesContext(ctx, cloudflare.WithPagination(p.PaginationOptions))
```

Comment:
You're only requesting the first page here. This line should be run in a loop, iterating through each page.

```go
page := 1

for {
  zonesResponse, err := p.Client.ListZonesContext(ctx, cloudflare.WithPagination(cloudflare.PaginationOptions{
    Page: page,
    PerPage: 50, // Always ask for the limit, since we're always going to iterate through the whole list.
  }))

  for _, zone := range zonesResponse.Result {
    ...work
  }

  if page == zonesResponse.ResultInfo.TotalPages {
    break
  }

  page++
}
```

Reply:
You could also optimize this query when a zone filter is provided as the `cloudflare-go` library provides [filtering](https://github.com/cloudflare/cloudflare-go/blob/0c85496d873009e53e64d391ade643e7ac02e0d4/cloudflare.go#L396).